Created a new guide for installing packages on AC (#286)
astronomer/docs: ac/next/install-dependencies.md

---
title: "Install Dependencies in a Production Environment"
navTitle: "Install Dependencies"
description: "Learn how to install OS-level and Python-level packages on Airflow."
---

## Overview

> By default, the Astronomer Core (AC) Airflow image includes several provider packages in order to better integrate between popular applications. For a full list of built in packages, read [Astronomer Core Image Architecture].

Reviewer: ```suggestion
By default, the Astronomer Core (AC) Docker image for Apache Airflow is built with a collection of pre-installed provider packages to help users integrate with popular applications. For the full list of built-in packages, read [Astronomer Core Image Architecture].
```

Reviewer: This intro is Docker-heavy - should this doc also apply to people using our python wheel? It's an important statement, but it might belong lower in the doc and be replaced with something more general

Author: I was using "image" here to describe both our VM offering and our Docker offering. What would be the appropriate word to describe both things at once? Distribution?

If you require additional dependencies to run your DAGs, you can either install those packages onto your local machines or build them into a Docker image.

## Install Packages on a Virtual Machine

To build Python and OS-level packages into a machine running Airflow, run the following command:

```sh

> sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-certified[<your-dependency>]==1.10.10.*'

Reviewer: ```suggestion
sudo -u astro ~astro/airflow-venv/bin/pip install --extra-index-url=https://pip.astronomer.io/simple/ 'astronomer-core[<your-dependency>]==<airflow-version>.*'
```

I think we standardize a comment around this command in another doc?

```

> You can also create your own Python packages and install them on your environment via Python wheel, or you can configure an environment variable to automatically add a Python package to your Airflow project folder. For more information on this setup, read the [Apache Airflow documentation on managing modules](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/modules_management.html).

Reviewer: ```suggestion
If you have your own Python packages, you can install them into your Airflow environment either via a Python wheel, or by configuring an environment variable to automatically add them  to your Airflow project directory. For more information on this setup, read the [Apache Airflow documentation on managing modules](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/modules_management.html).
```

Also, I thinkkk this isn't the right docs link - these look for dev docs. Should it be https://airflow.apache.org/docs/apache-airflow/stable/modules_management.html?

Author: Oh weird, I don't know why it has that weird host name. Fixed

## Install Packages Using the Astronomer CLI

> If you installed AC via the [Quickstart], you can add Python-level packages to your `requirements.txt` file and OS-level packages to your `packages.txt` file.

Reviewer: ```suggestion
If you're running Astronomer Core locally via the Astronomer CLI [Quickstart], you can add Python-level packages to your `requirements.txt` file and OS-level packages to your `packages.txt` file.
```

Author: Are we turning our existing Quickstart into a CLI quickstart? I was assuming this would link out to the AC Quickstart.

> To add the package, simply list the package name on its own line in one of the two files.
> 
> You can also pin a specific package version to use. For example, if wanted to exclusively use Pymongo 3.7.2, you'd add the following line to your `requirements.txt` file:

Reviewer: ```suggestion
To add the package, simply list the package name on its own line in the corresponding file. To install a particular version of any packages, pin that version to the package name in the format of `<package-name>==<package-version>`. To install Pymongo 3.7.2, for example, add the following to your `requirements.txt` file:
```

```
pymongo==3.7.2
```

If you don't pin a package to a version, the latest version of the package that's publicly available will be installed by default.

Once you've saved those packages to the appropriate `.txt` files, rebuild your image by running `astro dev stop`, followed by `astro dev start`. This process stops your running Docker containers and restarts them with your updated image.

## Install Packages on a Custom Docker Image

> If you [installed AC on Docker], you can add packages directly to your image. To install OS-level packages, you can specify them using a `RUN` directive with `apt-get`. For example, the following Dockerfile will install `your-dependency` on the image:

Reviewer: Is this specified in the user's Dockerfile?

Author: Yep, made that more clear in the first sentence.

```
FROM: quay.io/astronomer/docker-airflow:latest-onbuild
RUN apt-get update \
  && apt-get install -y --no-install-recommends \
         your-dependency \
  && apt-get autoremove -yqq --purge \
  && apt-get clean \
  && rm -rf /var/lib/apt/lists/*
```

> To install a Python-level package, specify the package using a `RUN` directive with `pip install` instead. For example:

Reviewer: Could we consider doing two H3s in this section, one for `### OS-level Packages` and another for `### Python Packages`? Don't feel strongly - just an idea.

Author: I feel like there's not enough content here to warrant splitting into H3s, and it's my guess that anyone seeking out this info would be likely to want to install both types of packages at once.

```
FROM: quay.io/astronomer/docker-airflow:latest-onbuild
RUN pip install --no-cache-dir --user your-dependency
```

Once you rebuild your image with `docker-build`, the image will have access to any packages that you specified. You can confirm that a package was installed by running a `$ docker exec` command into your Scheduler. To do so:

1. Run `$ docker ps` and retrieve the container ID of your Scheduler container.
2. Run the following command:

    ```
    docker exec -it <scheduler-container-id> pip freeze | grep <package-name>
    ```

    If the package was successfully installed, you should see the following output:

    ```
    <package-name>==<version>

>     ```

Reviewer: I _think_ this process for verifying that a package was successfully installed also applies to containers spun up via the Astro CLI? Looks like we mention it here: https://www.astronomer.io/docs/cloud/stable/develop/customize-image#add-python-and-os-level-packages

Author: True, I was struggling with how to connect this to both sections. Seeing how small the steps are now, I think I can simplify this and just add these 2 steps to both sections. 

Side note: do we need instructions like this for the VM setup? I was under the impression that successfully running `sudo -u astro...` would confirm installation on its own.
